openedx2zim/scraper.py

```diff
@@ -363,12 +388,174 @@ def download(self, connection):
+            meta.update({"optimizer_version": OPTIMIZER_VERSIONS[fpath.suffix[1:]]})
+        try:
+            self.s3_storage.upload_file(fpath, key, meta=meta)
+        except Exception as exc:
+            logger.error(f"{key} failed to upload to cache: {exc}")
+            return False
+        logger.info(f"uploaded {fpath} to cache at {key}")
+        return True
+
+    def downlaod_form_url(self, url, fpath, filetype):
+        download_path = fpath
+        if (
+            filetype
+            and (fpath.suffix[1:] != filetype)
+            and not (filetype == "jpeg" and fpath.suffix[1:] == "jpg")
+        ):
+            download_path = pathlib.Path(
+                tempfile.NamedTemporaryFile(
+                    suffix=f".{filetype}", dir=fpath.parent, delete=False
+                ).name
+            )
+        try:
+            save_large_file(url, download_path)
+            return download_path
+        except Exception as exc:
+            logger.error(f"Error while running save_large_file(): {exc}")
+            os.unlink(download_path)
+            return None
+
+    def download_from_youtube(self, url, fpath):
+        audext, vidext = {"webm": ("webm", "webm"), "mp4": ("m4a", "mp4")}[
+            self.video_format
+        ]
+        options = {
+            "outtmpl": fpath,
+            "preferredcodec": self.video_format,
+            "format": f"best[ext={vidext}]/bestvideo[ext={vidext}]+bestaudio[ext={audext}]/best",
+            "retries": 20,
+            "fragment-retries": 50,
+        }
+        try:
+            with youtube_dl.YoutubeDL(options) as ydl:
+                ydl.download([url])
+                return fpath
+        except Exception as exc:
+            logger.error(f"Error while running youtube_dl: {exc}")
+            return None
+
+    def convert_video(self, src, dst):
+        if (src.suffix[1:] != self.video_format) or self.low_quality:
+            preset = VideoWebmLow() if self.video_format == "webm" else VideoMp4Low()
+            reencode(
+                src, dst, preset.to_ffmpeg_args(), delete_src=True, failsafe=False,
+            )
+
+    def optimize_image(self, src, dst, resize=None):
+        if resize:
+            resize_image(src, width=resize, allow_upscaling=True)
+        if src.suffix[1:] == "jpeg":
+            exec_cmd("jpegoptim --strip-all -m50 " + str(src), timeout=10)
+        elif src.suffix[1:] == "png":
+            exec_cmd(
+                "pngquant --verbose --nofs --force --ext=.png " + str(src), timeout=10
+            )
+            exec_cmd("advdef -q -z -4 -i 5  " + str(src), timeout=50)
+        elif src.suffix[1:] == "gif":
+            exec_cmd("gifsicle --batch -O3 -i " + str(src), timeout=10)
+        if not src.resolve() == dst.resolve():
+            shutil.move(src, dst)
+
+    def optimize_file(self, src, dst):
+        if src.suffix[1:] in VIDEO_FORMATS:
+            self.convert_video(src, dst)
+        else:
```

Comment:
You may want to use another `if` instead so images are clearly filtered here.

Reply:
done